### PR TITLE
Wip/altran/use ssl

### PR DESCRIPTION
**Pull request by altran01**

### CODE REVIEW

File: python/veles/util/helpers.py [hunk 1]
```diff
@@ -21,15 +31,73 @@ def prepare_auth_key(key):
...
+
+
+def generate_ssl_cert(cert_path, key_path):
+    key = crypto.PKey()
+    key.generate_key(crypto.TYPE_RSA, 2048)
+
+    cert = crypto.X509()
+    cert.get_subject().C = "PL"
+    cert.get_subject().O = "Codilime"
+    cert.get_subject().OU = "Codisec"
```
Comment (adamiwaniuk) [inline python/veles/util/helpers.py]: please change that to something else

File: python/veles/util/helpers.py [hunk 2]
```diff
@@ -21,15 +31,73 @@ def prepare_auth_key(key):
...
+    cert.set_serial_number(1000)
+    cert.gmtime_adj_notBefore(0)
+    cert.gmtime_adj_notAfter(10 * 365 * 24 * 60 * 60)
+    cert.set_issuer(cert.get_subject())
+    cert.set_pubkey(key)
+    cert.sign(key, 'sha1')
+
+    with open(cert_path, 'w') as f:
+        f.write(crypto.dump_certificate(crypto.FILETYPE_PEM, cert).decode())
+    with open(key_path, 'w') as f:
```
Comment (adamiwaniuk) [inline python/veles/util/helpers.py]: what is default permission of this file? if it is readable by all, maybe it is worth to change that?

File: python/veles/util/helpers.py [hunk 3]
```diff
@@ -21,15 +32,74 @@ def prepare_auth_key(key):
...
+
+    cert = crypto.X509()
+    cert.get_subject().C = "RE"
+    cert.get_subject().O = "VELES"
+    cert.set_serial_number(1000)
+    cert.gmtime_adj_notBefore(0)
+    cert.gmtime_adj_notAfter(10 * 365 * 24 * 60 * 60)
+    cert.set_issuer(cert.get_subject())
+    cert.set_pubkey(key)
+    cert.sign(key, 'sha1')
```
Comment (adamiwaniuk) [inline python/veles/util/helpers.py]: sha1 !

**Comment by adamiwaniuk** [comment]
odpalone z instalatora:

> Process of locally created server started.
> Waiting for a new server to start...
> Traceback (most recent call last):
>   File "srv.py", line 23, in <module>
>     from veles.server.proto import create_unix_server, create_tcp_server
>   File "C:\Program Files (x86)\Veles\veles-server\python\veles\server\proto.py", line 27, in <module>
>     from OpenSSL import crypto
>   File "C:\Program Files (x86)\Veles\veles-server\python\OpenSSL\__init__.py", line 8, in <module>
>     from OpenSSL import rand, crypto, SSL
>   File "C:\Program Files (x86)\Veles\veles-server\python\OpenSSL\rand.py", line 12, in <module>
>     from OpenSSL._util import (
>   File "C:\Program Files (x86)\Veles\veles-server\python\OpenSSL\_util.py", line 6, in <module>
>     from cryptography.hazmat.bindings.openssl.binding import Binding
>   File "C:\Program Files (x86)\Veles\veles-server\python\cryptography\hazmat\bindings\openssl\binding.py", line 13, in <module>
>     from cryptography.hazmat.bindings._openssl import ffi, lib
> ImportError: DLL load failed: %1 nie jest prawidłową aplikacją systemu Win32.
> Process of locally created server finished. Exit code: 1.

File: python/proxy [hunk 4]
```diff
@@ -19,42 +19,50 @@ import logging
...
 
 parser = argparse.ArgumentParser(
-    parents=[helpers.get_client_argparse(), helpers.get_logging_argparse()])
+    parents=[helpers.get_client_argparse(), helpers.get_logging_argparse()],
+    add_help=False)
 parser.add_argument(
-    'url', help='either UNIX:<socket_path> or <ip>:<tcp port> to listen on')
+    'url', help='URL that the server will listen on, in form of: '
+                'VELES[+SSL|+UNIX]://auth_key@([host]:port|path) '
+                'there are 3 supported schemas: VELES - TCP socket, '
```
Comment (mkow) [inline python/proxy]: `\nThere`

File: python/srv.py [hunk 5]
```diff
@@ -20,20 +20,27 @@
...
 
 parser = helpers.get_logging_argparse()
 parser.add_argument(
-    'url', help='either UNIX:<socket_path> or [<ip>:]<tcp port> to listen on')
-parser.add_argument(
-    'auth_key', help='hex-encoded up to 64bytes value that '
-                     'clients need to provide when connecting')
+    'url', help='URL that the server will listen on, in form of: '
+                'VELES[+SSL|+UNIX]://auth_key@([host]:port|path) '
+                'there are 3 supported schemas: VELES - TCP socket, '
```
Comment (mkow) [inline python/srv.py]: ditto

File: python/veles/server/proto.py [hunk 6]
```diff
@@ -19,8 +19,12 @@
 
 import asyncio
 import hmac
+import logging
+from os import path
+import ssl
```
Comment (mkow) [inline python/veles/server/proto.py]: sort please

File: python/veles/util/helpers.py [hunk 7]
```diff
@@ -21,15 +32,74 @@ def prepare_auth_key(key):
 
 def get_client_argparse():
     parser = argparse.ArgumentParser()
-    parser.add_argument('server', help='server address')
-    parser.add_argument('server_auth_key', help='hex-encoded server auth-key')
+    parser.add_argument(
+        'server_url',
+        help='server URL in form of: VELES[+SSL|+UNIX]://auth_key[:cert-'
+             'fingerprint]@([host]:port|path) '
+             'there are 3 supported schemas: VELES - TCP socket, '
```
Comment (mkow) [inline python/veles/util/helpers.py]: ditto

File: src/client/networkclient.cc [hunk 8]
```diff
@@ -99,9 +131,18 @@ void NetworkClient::connect(
...
-    client_socket_ = new QTcpSocket(this);
-    QObject::connect(client_socket_, &QAbstractSocket::connected,
-        this, &NetworkClient::socketConnected, Qt::QueuedConnection);
+    client_socket_ = new QSslSocket(this);
+    if (ssl_enabled_) {
+      QObject::connect(client_socket_, &QSslSocket::encrypted,
+          this, &NetworkClient::socketConnected, Qt::QueuedConnection);
+      QObject::connect(
+          client_socket_,
+          static_cast<void(QSslSocket::*)(const QList<QSslError> &)>(&QSslSocket::sslErrors),
```
Comment (mkow) [inline src/client/networkclient.cc]: no space (`const QList<QSslError>&`)

File: src/client/networkclient.cc [hunk 9]
```diff
@@ -123,7 +164,11 @@ void NetworkClient::connect(
       if (output()) {
         *output() << "NetworkClient: bind successful." << endl;
       }
-      client_socket_->connectToHost(server_name_, server_port_);
+      if(ssl_enabled_) {
```
Comment (mkow) [inline src/client/networkclient.cc]: space

File: src/client/networkclient.cc [hunk 10]
```diff
@@ -459,5 +504,40 @@ void NetworkClient::socketError(QAbstractSocket::SocketError socketError) {
   }
 }
 
+void NetworkClient::checkFingerprint(const QList<QSslError>& errors) {
+  bool fingerprint_valid = false;
+  for (auto err : errors) {
```
Comment (mkow) [inline src/client/networkclient.cc]: const ref

File: src/ui/connectiondialog.cc [hunk 11]
```diff
@@ -72,6 +74,15 @@ ConnectionDialog::ConnectionDialog(QWidget* parent)
       server_file_dialog_, &QFileDialog::show);
   connect(server_file_dialog_, &QFileDialog::fileSelected,
       this, &ConnectionDialog::serverFileSelected);
+
+  certificate_dir_dialog_ = new QFileDialog(this);
+  certificate_dir_dialog_->setFileMode(QFileDialog::Directory);
+
+  connect(ui_->select_database_button, &QPushButton::clicked,
+        certificate_dir_dialog_, &QFileDialog::show);
```
Comment (mkow) [inline src/ui/connectiondialog.cc]: fix indent

File: src/ui/connectiondialog.cc [hunk 12]
```diff
@@ -80,8 +91,34 @@ ConnectionDialog::ConnectionDialog(QWidget* parent)
   connect(ui_->remove_profile_button, &QPushButton::clicked, this, &ConnectionDialog::profileRemoved);
   connect(ui_->new_profile_button, &QPushButton::clicked, this, &ConnectionDialog::newProfile);
   connect(ui_->default_profile, &QPushButton::clicked, this, &ConnectionDialog::defaultProfile);
+  connect(ui_->ssl_checkbox, &QCheckBox::toggled, this, &ConnectionDialog::sslEnabledToggled);
+  connect(ui_->server_url_line_edit, &QLineEdit::textEdited, [this](const QString &text) {
```
Comment (mkow) [inline src/ui/connectiondialog.cc]: `&` to the left

File: src/ui/connectiondialog.cc [hunk 13]
```diff
@@ -122,6 +159,18 @@ QString ConnectionDialog::serverScript() {
   return ui_->server_executable_line_edit->text();
 }
 
+QString ConnectionDialog::certificateDir() {
```
Comment (mkow) [inline src/ui/connectiondialog.cc]: should be const

File: src/ui/connectiondialog.cc [hunk 14]
```diff
@@ -122,6 +159,18 @@ QString ConnectionDialog::serverScript() {
...
+  return ui_->certificate_dir_line_edit->text();
+}
+
+QString ConnectionDialog::serverUrl() {
```
Comment (mkow) [inline src/ui/connectiondialog.cc]: ditto

File: src/ui/connectiondialog.cc [hunk 15]
```diff
@@ -122,6 +159,18 @@ QString ConnectionDialog::serverScript() {
...
+  return ui_->server_url_line_edit->text();
+}
+
+bool ConnectionDialog::sslEnabled() {
```
Comment (mkow) [inline src/ui/connectiondialog.cc]: ditto

File: src/ui/connectiondialog.cc [hunk 16]
```diff
@@ -196,12 +276,21 @@ void ConnectionDialog::loadProfiles() {
...
+    ui_->port_spin_box->setValue(util::settings::connection::serverPort());
+    ui_->key_line_edit->setText(util::settings::connection::connectionKey());
+    ui_->ssl_checkbox->setChecked(
+        util::settings::connection::sslEnabled());
+  } else {
+    ui_->existing_server_radio_button->setChecked(true);
+    ui_->server_url_line_edit->setText(
+        util::settings::connection::serverUrl());
+    emit ui_->server_url_line_edit->textEdited(
+          util::settings::connection::serverUrl());
```
Comment (mkow) [inline src/ui/connectiondialog.cc]: fix indent

File: src/ui/connectiondialog.ui [hunk 17]
```diff
@@ -200,90 +200,35 @@
...
        <property name="text">
-        <string>Client name:</string>
-       </property>
-      </widget>
-     </item>
-     <item row="4" column="1">
-      <widget class="QLineEdit" name="client_name_line_edit">
-       <property name="placeholderText">
-        <string>Client name visible for other users</string>
+        <string>Certificate diectory:</string>
```
Comment (mkow) [inline src/ui/connectiondialog.ui]: `directory`

File: src/ui/connectionmanager.cc [hunk 18]
```diff
@@ -234,11 +239,33 @@ void ConnectionManager::disconnect() {
 }
 
 void ConnectionManager::serverProcessReadyRead() {
-  char buf[10240];
   if(server_process_ && server_process_->bytesAvailable() > 0) {
-    qint64 len = server_process_->read(buf, sizeof(buf));
-    if(len > 0) {
-      LogWidget::output()->write(buf, len);
+    QByteArray out = server_process_->read(server_process_->bytesAvailable());
+    if(out.size() > 0) {
```
Comment (mkow) [inline src/ui/connectionmanager.cc]: space

File: src/util/settings/connection_client.cc [hunk 19]
```diff
@@ -235,6 +235,45 @@ void setServerScript(QString server_script) {
...
+QString serverUrlDefault() {
+  return "";
+}
+
+QString serverUrl() {
+  return profileSettings("connection.server_url",
+      serverUrlDefault()).toString();
+}
+
+void setServerUrl(QString server_url) {
```
Comment (mkow) [inline src/util/settings/connection_client.cc]: const ref

**Pull request merged**